codelists/presenters.py

```diff
@@ -1,8 +1,72 @@
 from itertools import groupby
 
+import attr
 from django.urls import reverse
 
 
+@attr.s
+class Row:
+    """
+    Data structure for a Definition to prepare for display
+
+    name: name of the definition
+    code: code of the definition
+    excluded_children: list of excluded children (created with this class too)
+    all_descendants: are all of this Definitions descendants included?
+
+    Between all_descendants and excluded_children we can cover the three state
+    situation a Definition can be in:
+
+        * all descendants are included (all_descendants = True, excluded_children = [])
+        * all descendants except N (all_descendants = True, excluded_children = [...])
+        * no descendants (all_descendants = False, excluded_children ignored)
+
+    """
+
+    name: str = attr.ib()
+    code: str = attr.ib()
+    excluded_children: list = attr.ib(default=list())
+    all_descendants: bool = attr.ib(default=True)
+
+
+def _iter_rules(hierarchy, rules, name_for_rule, excluded=None):
```

Comment:
This works while a `DefinitionRule` does not have both `applies_to_descendants=True` and `code_is_excluded=True`.  That is currently the case because of how `Definition.from_codes` works.

If this wasn't the case, we'd end up recursing a second time (ie calling the function three times) which would mean that `r in excluded` raises a `TypeError` because `excluded` is `None`.

I wonder whether it's worth adding a check (an `assert` would do) to `DefinitionRule.__init__` to ensure that the the invariant is maintained.

Reply:
Makes sense!  I've added it in c44bbcb as an exception so I can express why it happened and for what code (which should help traceability when we hit this $later).

Reply:
Great.